rsocket-core/src/main/java/io/rsocket/fragmentation/FrameReassembler.java

```diff
@@ -25,143 +37,252 @@
-      data = accumulateData(fragment);
+    if (FrameHeaderFlyweight.hasMetadata(frame)) {
+      CompositeByteBuf metadata = getMetadata(streamId);
+      switch (frameType) {
+        case REQUEST_FNF:
+          metadata.addComponents(true, RequestFireAndForgetFrameFlyweight.metadata(frame));
+          break;
+        case REQUEST_STREAM:
+          metadata.addComponents(true, RequestStreamFrameFlyweight.metadata(frame));
+          break;
+        case REQUEST_RESPONSE:
+          metadata.addComponents(true, RequestResponseFrameFlyweight.metadata(frame));
+          break;
+        case REQUEST_CHANNEL:
+          metadata.addComponents(true, RequestChannelFrameFlyweight.metadata(frame));
+          break;
+          // Payload and synthetic types
+        case PAYLOAD:
+        case NEXT:
+        case NEXT_COMPLETE:
+        case COMPLETE:
+          metadata.addComponents(true, PayloadFrameFlyweight.metadata(frame));
+          break;
+        default:
+          throw new IllegalStateException("unsupported fragment type");
+      }
     }
 
-    Frame createFrame(ByteBufAllocator byteBufAllocator) {
-      FragmentableFrame root = fragments.get(0);
-      return root.createNonFragment(byteBufAllocator, metadata, data);
+    ByteBuf data;
+    switch (frameType) {
+      case REQUEST_FNF:
+        data = RequestFireAndForgetFrameFlyweight.data(frame);
+        break;
+      case REQUEST_STREAM:
+        data = RequestStreamFrameFlyweight.data(frame);
+        break;
+      case REQUEST_RESPONSE:
+        data = RequestResponseFrameFlyweight.data(frame);
+        break;
+      case REQUEST_CHANNEL:
+        data = RequestChannelFrameFlyweight.data(frame);
+        break;
+        // Payload and synthetic types
+      case PAYLOAD:
+      case NEXT:
+      case NEXT_COMPLETE:
+      case COMPLETE:
+        data = PayloadFrameFlyweight.data(frame);
+        break;
+      default:
+        throw new IllegalStateException("unsupported fragment type");
     }
 
-    private ByteBuf accumulateData(FragmentableFrame fragment) {
-      ByteBuf data = fragment.getUnsafeData();
-      return this.data == null ? data.retain() : Unpooled.wrappedBuffer(this.data, data.retain());
+    if (data != Unpooled.EMPTY_BUFFER) {
+      getData(streamId).addComponents(true, data);
     }
+  }
+
+  void reassembleFrame(ByteBuf frame, SynchronousSink<ByteBuf> sink) {
+    try {
+      FrameType frameType = FrameHeaderFlyweight.frameType(frame);
+      int streamId = FrameHeaderFlyweight.streamId(frame);
+      switch (frameType) {
+        case CANCEL:
+        case ERROR:
+          cancelAssemble(streamId);
+        default:
+      }
 
-    private @Nullable ByteBuf accumulateMetadata(FragmentableFrame fragment) {
-      ByteBuf metadata = fragment.getUnsafeMetadata();
+      if (!frameType.isFragmentable()) {
+        sink.next(frame);
+        return;
+      }
+
+      boolean hasFollows = FrameHeaderFlyweight.hasFollows(frame);
```

Comment:
nit: there is a mention in a spec that `follows` flag should be ignored if `complete` flag is set